File: contracts/wasm/wasmlib/ts/wasmlib/contract.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// base contract objects

import {BytesEncoder} from "./bytes";
import {ROOT,ScTransfers} from "./context";
import {ScChainID,ScHname} from "./hashtypes";
import {getObjectID, panic, TYPE_MAP} from "./host";
import * as keys from "./keys";
import {ScMutableMap} from "./mutable";

export interface ScFuncCallContext {
    canCallFunc():void;
}

export interface ScViewCallContext {
    canCallView():void;
}

export class ScMapID {
    mapID: i32 = 0;
}

type NullableScMapID = ScMapID | null;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScView {
    hContract: ScHname;
    hFunction: ScHname;
    paramsID: NullableScMapID;
    resultsID: NullableScMapID;

    constructor(hContract: ScHname, hFunction: ScHname) {
        this.hContract = hContract;
        this.hFunction = hFunction;
        this.paramsID = null;
        this.resultsID = null;
    }

    setPtrs(paramsID: NullableScMapID, resultsID: NullableScMapID): void {
        this.paramsID = paramsID;
        this.resultsID = resultsID;

        if (paramsID === null) {
        } else {
            paramsID.mapID = ScMutableMap.create().mapID();
        }
    }

    call(): void {
        this.callWithTransfer(0);
    }

    callWithTransfer(transferID: i32): void {
        let encode = new BytesEncoder();
        encode.hname(this.hContract);
        encode.hname(this.hFunction);
        encode.int32(this.id(this.paramsID));
        encode.int32(transferID);
        ROOT.getBytes(keys.KEY_CALL).setValue(encode.data());

        let resultsID = this.resultsID;
        if (resultsID === null) {
        } else {
            resultsID.mapID = getObjectID(1, keys.KEY_RETURN, TYPE_MAP);
        }
    }

    clone(): ScView {
        let o = new ScView(this.hContract, this.hFunction);
        o.paramsID = this.paramsID;
        o.resultsID = this.resultsID;
        return o;
    }

    ofContract(hContract: ScHname): ScView {
        let ret = this.clone();
        ret.hContract = hContract;
        return ret;
    }

    id(paramsID: NullableScMapID): i32 {
        if (paramsID === null) {
            return 0;
        }
        return paramsID.mapID;
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScInitFunc {
    view: ScView;

    constructor(hContract: ScHname, hFunction: ScHname) {
        this.view = new ScView(hContract, hFunction);
    }

    setPtrs(paramsID: NullableScMapID, resultsID: NullableScMapID): void {
        this.view.setPtrs(paramsID, resultsID);
    }

    call(): void {
        return panic("cannot call init");
    }

    clone(): ScInitFunc {
        let o = new ScInitFunc(this.view.hContract, this.view.hFunction);
        o.view = this.view.clone();
        return o;
    }

    ofContract(hContract: ScHname): ScInitFunc {
        let ret = this.clone();
        ret.view.hContract = hContract;
        return ret;
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScFunc {
    view: ScView;
    delaySeconds: i32 = 0;
    transferID: i32 = 0;

    constructor(hContract: ScHname, hFunction: ScHname) {
        this.view = new ScView(hContract, hFunction);
    }

    setPtrs(paramsID: NullableScMapID, resultsID: NullableScMapID): void {
        this.view.setPtrs(paramsID, resultsID);
    }

    call(): void {
        if (this.delaySeconds != 0) {
            return panic("cannot delay a call");
        }
        this.view.callWithTransfer(this.transferID);
    }

    clone(): ScFunc {
        let o = new ScFunc(this.view.hContract, this.view.hFunction);
        o.view = this.view.clone();
        o.delaySeconds = this.delaySeconds;
        o.transferID = this.transferID;
        return o;
    }

    delay(seconds: i32): ScFunc {
        let ret = this.clone();
        ret.delaySeconds = seconds;
        return ret;
    }

    ofContract(hContract: ScHname): ScFunc {
        let ret = this.clone();
        ret.view.hContract = hContract;
        return ret;
    }

    post(): void {
        return this.postToChain(ROOT.getChainID(keys.KEY_CHAIN_ID).value());
    }

    postToChain(chainID: ScChainID): void {
        let encode = new BytesEncoder();
        encode.chainID(chainID);
        encode.hname(this.view.hContract);
        encode.hname(this.view.hFunction);
        encode.int32(this.view.id(this.view.paramsID));
        encode.int32(this.transferID);
        encode.int32(this.delaySeconds);
        ROOT.getBytes(keys.KEY_POST).setValue(encode.data());
    }

    transfer(transfer: ScTransfers): ScFunc {
        let ret = this.clone();
        ret.transferID = transfer.transfers.objID;
        return ret;
    }

    transferIotas(amount: i64): ScFunc {
        return this.transfer(ScTransfers.iotas(amount));
    }
}

File: packages/vm/wasmlib/src/contract.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// base contract objects

use std::ptr;

use crate::bytes::*;
use crate::context::*;
use crate::hashtypes::*;
use crate::host::*;
use crate::keys::*;
use crate::mutable::*;

pub trait ScFuncCallContext {
    fn can_call_func(&self);
}

pub trait ScViewCallContext {
    fn can_call_view(&self);
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

#[derive(Clone, Copy)]
pub struct ScView {
    h_contract: ScHname,
    h_function: ScHname,
    params_id: *mut i32,
    results_id: *mut i32,
}

impl ScView {
    pub fn new(h_contract: ScHname, h_function: ScHname) -> ScView {
        ScView {
            h_contract: h_contract,
            h_function: h_function,
            params_id: ptr::null_mut(),
            results_id: ptr::null_mut(),
        }
    }

    pub fn set_ptrs(&mut self, params_id: *mut i32, results_id: *mut i32) {
        self.params_id = params_id;
        self.results_id = results_id;

        unsafe {
            if params_id != ptr::null_mut() {
                *params_id = ScMutableMap::new().map_id();
            }
        }
    }

    pub fn call(&self) {
        self.call_with_transfer(0);
    }

    fn call_with_transfer(&self, transfer_id: i32) {
        let mut encode = BytesEncoder::new();
        encode.hname(&self.h_contract);
        encode.hname(&self.h_function);
        encode.int32(self.id(self.params_id));
        encode.int32(transfer_id);
        ROOT.get_bytes(&KEY_CALL).set_value(&encode.data());

        unsafe {
            if self.results_id != ptr::null_mut() {
                *self.results_id = get_object_id(1, KEY_RETURN, TYPE_MAP);
            }
        }
    }

    pub fn of_contract(&self, h_contract: ScHname) -> ScView {
        let mut ret = self.clone();
        ret.h_contract = h_contract;
        ret
    }

    fn id(&self, params_id: *mut i32) -> i32 {
        unsafe {
            if params_id == ptr::null_mut() {
                return 0;
            }
            *params_id
        }
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

#[derive(Clone, Copy)]
pub struct ScInitFunc {
    view: ScView,
}

impl ScInitFunc {
    pub fn new(h_contract: ScHname, h_function: ScHname) -> ScInitFunc {
        ScInitFunc {
            view: ScView::new(h_contract, h_function),
        }
    }

    pub fn set_ptrs(&mut self, params_id: *mut i32, results_id: *mut i32) {
        self.view.set_ptrs(params_id, results_id);
    }

    pub fn call(&self) {
        panic("cannot call init")
    }

    pub fn of_contract(&self, h_contract: ScHname) -> ScInitFunc {
        let mut ret = self.clone();
        ret.view.h_contract = h_contract;
        ret
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

#[derive(Clone, Copy)]
pub struct ScFunc {
    view: ScView,
    delay: i32,
    transfer_id: i32,
}

impl ScFunc {
    pub fn new(h_contract: ScHname, h_function: ScHname) -> ScFunc {
        ScFunc {
            view: ScView::new(h_contract, h_function),
            delay: 0,
            transfer_id: 0,
        }
    }

    pub fn set_ptrs(&mut self, params_id: *mut i32, results_id: *mut i32) {
        self.view.set_ptrs(params_id, results_id);
    }

    pub fn call(&self) {
        if self.delay != 0 {
            panic("cannot delay a call")
        }
        self.view.call_with_transfer(self.transfer_id);
    }

    pub fn delay(&self, seconds: i32) -> ScFunc {
        let mut ret = self.clone();
        ret.delay = seconds;
        ret
    }

    pub fn of_contract(&self, h_contract: ScHname) -> ScFunc {
        let mut ret = self.clone();
        ret.view.h_contract = h_contract;
        ret
    }

    pub fn post(&self) {
        self.post_to_chain(ROOT.get_chain_id(&KEY_CHAIN_ID).value())
    }

    pub fn post_to_chain(&self, chain_id: ScChainID) {
        let mut encode = BytesEncoder::new();
        encode.chain_id(&chain_id);
        encode.hname(&self.view.h_contract);
        encode.hname(&self.view.h_function);
        encode.int32(self.view.id(self.view.params_id));
        encode.int32(self.transfer_id);
        encode.int32(self.delay);
        ROOT.get_bytes(&KEY_POST).set_value(&encode.data());
    }

    pub fn transfer(&self, transfer: ScTransfers) -> ScFunc {
        let mut ret = self.clone();
        ret.transfer_id = transfer.transfers.obj_id;
        ret
    }

    pub fn transfer_iotas(&self, amount: i64) -> ScFunc {
        self.transfer(ScTransfers::iotas(amount))
    }
}
